File: src/148-排序链表/index2.js
/**
 * @description 148-排序链表
 * @author 氧化氢
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 * @solution 这种写法是通过递归然后归并
 */
var sortList = function (head) {
  if (!head || !head.next) return head
  // 利用快慢指针知道中间节点，然后将其断开
  var slow = head
  var fast = head.next
  while (fast && fast.next) {
    slow = slow.next
    fast = fast.next.next
  }
  var right = slow.next
  slow.next = null
  var left = head

  // 递归排序：
  var p = sortList(left)
  var q = sortList(right)
  // 合并两个节点：
  var hair = new ListNode(-1),
    cur = hair
  while (p && q) {
    if (p.val >= q.val) {
      cur.next = q
      q = q.next
    } else {
      cur.next = p
      p = p.next
    }
    cur = cur.next
  }
  cur.next = p ? p : q

  return hair.next
};

function ListNode(val, next) {
  this.val = (val === undefined ? 0 : val)
  this.next = (next === undefined ? null : next)
}

export default sortList

File: src/148-排序链表/index2.py
#
# @lc app=leetcode.cn id=148 lang=python3
#
# [148] 排序链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        使用 n ** 2 的时间复杂度进行排序
        """
        if not head or not head.next:
            return head
        # 将链表分为 左、右 两个链表
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        right = slow.next
        slow.next = None
        left = head

        l = self.sortList(left)
        r = self.sortList(right)

        # 将两个排序后的链表进行 排序合并 后返回
        hair = ListNode(-1)
        cur = hair
        while l and r:
            if l.val <= r.val:
                cur.next = l
                l = l.next
            else:
                cur.next = r
                r = r.next
            cur = cur.next
        cur.next = l if l else r

        return hair.next


# @lc code=end
